### PR TITLE
Translate Math.PI

### PR DESCRIPTION
**Pull request by Marusyk**
Issue #25049

I've added a new member translator. I expected that it will run after this test
```
[ConditionalTheory]
[MemberData(nameof(IsAsyncData))]
public virtual Task Where_math_pi(bool async)
{
            return AssertQuery(
                async,
                ss => ss.Set<Product>()
                    .Where(od => Math.Abs(od.ProductID) > Math.PI),
                entryCount: 0);
}
```
Could you please suggest what I am doing wrong?
It just translates `Math.PI` to constant without calling my translator. It seems I missed something  🤔

### CODE REVIEW

File: src/EFCore.SqlServer/Query/Internal/SqlServerMathPIMemberTranslator.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            MemberInfo member,
+            Type returnType,
+            IDiagnosticsLogger<DbLoggerCategory.Query> logger)
+        {
+            Check.NotNull(member, nameof(member));
+            Check.NotNull(returnType, nameof(returnType));
+            Check.NotNull(logger, nameof(logger));
+
+            if (member.Name == nameof(Math.PI)
+                && instance?.Type == typeof(double))
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerMathPIMemberTranslator.cs:51]: `Math.PI` is a static member access. So the instance is always null hence this check fails.
Correct check (apart from matching member.Name) would be `member.DeclaringType == typeof(Math)`

Comment (smitpatel) [reply src/EFCore.SqlServer/Query/Internal/SqlServerMathPIMemberTranslator.cs:51]: (also check instance is null)

**Comment by smitpatel** [comment]
> It just translates Math.PI to constant without calling my translator.

You will also need to mark it as non-evaluatable in parameter extraction. https://github.com/dotnet/efcore/blob/main/src/EFCore.SqlServer/Query/Internal/SqlServerEvaluatableExpressionFilter.cs
In this file return false for Math.PI expression (using usual pattern matching)

File: src/EFCore.SqlServer/Query/Internal/SqlServerMathPIMemberTranslator.cs (hunk 1 above)
Comment (Marusyk) [reply src/EFCore.SqlServer/Query/Internal/SqlServerMathPIMemberTranslator.cs:51]: yeah, I will fix that but first I need to run it.

**Comment by Marusyk** [comment]
@smitpatel thanks for advice
The `IsEvaluatableExpression` catches the following 
`{[Microsoft.EntityFrameworkCore.Query.QueryRootExpression].Where(od => (3,141592653589793 > Convert(od.ProductID, Double)))}`
for test ` ss => ss.Set<Product>().Where(od => Math.PI > od.ProductID)`
Could you please suggest how to return false for Math.PI expression?

**Comment by smitpatel** [comment]
With that expression tree, we cannot add translation. That expression tree indicates that LINQ compiler inline the constant value when generating expression tree which happens even before the tree reaches EF Core.

cc: @bricelam

**Comment by bricelam** [comment]
Lame. Probably not worth translating, then.

**Comment by smitpatel** [comment]
Closing the PR. Will add notes on the issue. @Marusyk - Thanks for investigating this.

**Pull request closed without merging**